massa-async-pool/src/pool.rs

```diff
@@ -0,0 +1,178 @@
+    /// * excess messages after inserting all remaining new_messages, in priority order (from lowest to highest priority)
+    pub fn settle_slot(
+        &mut self,
+        slot: Slot,
+        new_messages: Vec<AsyncMessage>,
+    ) -> Vec<(AsyncMessageId, AsyncMessage)> {
+        // Compute IDs
+        let mut new_messages: Vec<_> = new_messages
+            .into_iter()
+            .map(|v| (v.compute_id(), v))
+            .collect();
+
+        // Filter out all messages for which the validity end is expired.
+        // Note that the validity_end bound is NOT included in the validity interval of the message.
+        let mut eliminated: Vec<_> = self
+            .messages
+            .drain_filter(|_k, v| slot >= v.validity_end)
+            .chain(new_messages.drain_filter(|(_k, v)| slot > v.validity_end))
+            .collect();
+
+        // Insert new messages into the pool
+        self.messages.extend(new_messages);
+
+        // Truncate message pool to its max size, removing non-prioritary items
+        let excess_count = self
+            .messages
+            .len()
+            .saturating_sub(self.config.max_length as usize);
+        eliminated.reserve_exact(excess_count);
+        for _ in 0..excess_count {
+            eliminated.push(self.messages.pop_first().unwrap()); // will not panic (checked at excess_count computation)
+        }
+
+        eliminated
+    }
+
+    /// Takes the best possible batch of messages to execute, with gas limits and slot validity filtering.
+    /// The returned messages are removed from the pool.
+    /// This method is used at the beginning of a slot execution to list async messages to execute.
+    ///
+    /// # arguments
+    /// * slot: select only messages that are valid within this slot
+    /// * available_gas: maximum amount of available gas
+    ///
+    /// # returns
+    /// A vector of messages, sorted from the most prioritary to the least prioritary
+    pub fn take_batch_to_executte(
+        &mut self,
+        slot: Slot,
+        mut available_gas: u64,
+    ) -> Vec<AsyncMessage> {
+        let mut selected = Vec::new();
+
+        // iterate in decreasing priority order
+        for (msg_id, msg) in self.messages.iter().rev() {
+            // check validity period
+            if slot < msg.validity_start || slot >= msg.validity_end {
+                continue;
+            }
+
+            // check available gas
+            if available_gas < msg.max_gas {
+                continue;
+            }
+
+            // add to selected items
+            selected.push(*msg_id);
+
+            // substract available gas
+            available_gas -= msg.max_gas;
+
+            // if there is no more gas, quit
+            if available_gas == 0 {
+                break;
+            }
+        }
+
+        // gather all selected items and remove them from self.messages
+        let mut accumulator = Vec::with_capacity(selected.len());
```

Comment:
I think this is an unnecessary loop. As you already filter them above I think you can use `drain_filter` (Example: https://doc.rust-lang.org/std/vec/struct.Vec.html#examples-47) above so that you directly remove them from `messages` and have them in a separate vector without this loop.